framework/i18n/Formatter.php

```diff
@@ -1741,4 +1778,220 @@ protected function createNumberFormatter($style, $decimals = null, $options = []
+            }
+
+            $decimalOutput .= $decimalSeparator . $fractionalPart;
+        }
+
+        if ($roundIntegerPart || ($decimals === 0 && $fractionalPart !== null && (int) substr($fractionalPart, 0, 1) >= 5)) {
+            $cursor = 0;
+            while (++$cursor <= strlen($integerPart)) {
+                $round = false;
+                $newDigit = (int) substr($integerPart, -$cursor, 1) + 1;
+
+                if ($newDigit === 10) {
+                    $newDigit = 0;
+                    $round = true;
+                }
+
+                $integerPart = substr($integerPart, 0, -$cursor) . $newDigit . substr($integerPart, strlen($integerPart) - $cursor + 1);
+
+                if (!$round) {
+                    break;
+                }
+            }
+            if ($integerPart === str_repeat('0', strlen($integerPart))) {
+                $integerPart = '1' . $integerPart;
+            }
+        }
+
+        if (strlen($integerPart) > 3) {
+            $thousandSeparator = $this->thousandSeparator;
+            if ($thousandSeparator === null) {
+                $thousandSeparator = ',';
+            }
+
+            $integerPart = strrev(implode(',', str_split(strrev($integerPart), 3)));
+            if ($thousandSeparator !== ',') {
+                $integerPart = str_replace(',', $thousandSeparator, $integerPart);
+            }
+        }
+
+        return $integerPart . $decimalOutput;
+    }
+
+    /**
+     * Fallback for formatting value as an integer number by removing any decimal digits without rounding.
+     *
+     * @param string|int|float $value the value to be formatted.
+     * @return string the formatted result.
+     * @since 2.0.16
+     */
+    protected function asIntegerStringFallback($value)
+    {
+        if (empty($value)) {
+            $value = 0;
+        }
+
+        $value = (string) $value;
+
+        if (strpos($value, '.') !== false) {
+            $integerPart = substr($value, 0, strrpos($value, '.'));
+        } else {
+            $integerPart = $value;
+        }
+
+        return $this->asDecimalStringFallback($integerPart, 0);
+    }
+
+    /**
+     * Fallback for formatting value as a percent number with "%" sign.
+     *
+     * Property [[decimalSeparator]] will be used to represent the decimal point. The value is rounded automatically
+     * to the defined decimal digits.
+     *
+     * @param string|int|float $value the value to be formatted.
+     * @param int $decimals the number of digits after the decimal point. The default value is `0`.
+     * @return string the formatted result.
+     * @since 2.0.16
+     */
+    protected function asPercentStringFallback($value, $decimals = null)
+    {
+        if (empty($value)) {
+            $value = 0;
+        }
+
+        if ($decimals === null) {
+            $decimals = 0;
+        }
+
+        $value = (string) $value;
+
+        if (strpos($value, '.') !== false) {
```

Comment:
`strpos($value, '.')` should be extracted in variable not to call it multiple times.